content/blog/2025-05-13-abstract-testing-floating-point.md

```diff
@@ -0,0 +1,201 @@
+error. To be precise, for every operation, the following overapproximation holds
+for some unit round-off value $u$:
+
+> $(op_{float} \ x \ y) = (op_{real} \ x \ y) * (1 + \epsilon)$
+
+for $|\epsilon| <= u$. Different tools may use varying overapproximations, but the
+principle is the same: to tractably verify you (typically) need to
+overapproximate.
+
+But how to test? Here's an idea: one can view an overapproximation as a bigger
+testing budget to smash the verification tool with. For floating-point
+verification, we can simply add $\epsilon$ error at every operation without
+worrying about whether the IEEE floating-point spec actually allows us to. In
+other words, the testing game is to find input $x$ and $\epsilon$-trace
+$\epsilon_1$, $\epsilon_2$, $\epsilon_3$ ... added at run-time such that we can violate
+**GUARANTEE**.
+
+That's my project, in a nutshell! [^1] 
+
+```
+
+                                           Abstract Testing
+                                              ======= 
+                                              =     =
+                                              =     =
+                                              =======
+                                                | |
+                                                | |
+                                                | |
+                                                | |
+                                                | |
+                                                |_|
+
+                                _______
+                               /       \
+                            ---|       |---
+                               |  Bug  |
+                            ---|       |---
+                               \_______/
+                                /  |  \
+                               /   |   \
+                              "    |    "
+```
+
+
+### Side quest: finding a good $\epsilon$-trace
+A natural question to ask is: how do we find a good $\epsilon$-trace? To build
+some intuition, consider a simple floating point program:
+
+$$p := \frac{a - b}{c + d}$$
+
+We wish to maximize the final error term in this program. Assuming that $a, b,
+c, d$ are all non-negative, it suffices to:
+
+1. Maximize the error term in the numerator ($a-b$). Recursing further, we
+   realize we wish to
+   - minimize the error term in $b$.
+   - maximize the error term in $a$, and,
+2. Minimize the error term in the denominator ($c + d$). Doing the obvious
+   thing, we realize we need to
+   - minimize the error term in $c$, and,
+   - minimize the error term in $d$.
+
+Once we know which direction to point each $\epsilon$, we can assign the most
+extreme values (either $+u$ or $-u$) to get a pretty good $\epsilon$-trace.
+
+For this program, the $\epsilon$-trace we computed happens to be the
+**worst-case**, which is awesome. This technique generalizes in a pretty
+straightforward manner to a backwards static analysis. [^2] 
```

Comment:
Does that static analysis _always_ produce the worst-case epsilon-trace? I'm assuming not… if not, can you explain a little bit more about when it is conservative (in the sense that there may exist an even worse/larger epsilon-trace)? Seeing one example of such a program would be instructive!

Reply:
Thanks for the suggestion! I added the following footnote to avoid breaking up the flow of the text too much:
```
+[^2]: A wrinkle occurs when terms can be shared. If a verification tool cannot
+    reason about shared error terms, this algorithm still produces a worst-case
+$\epsilon$-trace. Otherwise, this one might need to do a little searching to
+find the worst-case $\epsilon$-trace. In particular, if error terms can be
+shared, we start to care about the constants in the floating-point expression.
+Concretely, `(x - 2x)` is quite different from `(2x - x)`; in the former we wish
+to minimize the error term in `x`, while in the latter we wish to maximize the
+error term in `x`.
```